File: sdks/node/src/seed/resources/mesh.ts
/**
 * Mesh observability resource (ADR-0016a §D8 Phase 1 surface addendum).
 *
 * Four read-only seed endpoints that make the mesh observable from an
 * SDK consumer. All four are WiFi-read allowlist endpoints — no pairing
 * token required (see `seed/src/cognitum-agent/src/api.rs:392-400`).
 *
 * Wire shapes verified against live seed v0.20.0 on 2026-04-22 — see
 * `../models/mesh.ts` for JSON samples.
 */

import type { CallOptions } from "../callOptions.js";
import type {
  ClusterHealth,
  MeshPeers,
  MeshStatus,
  SwarmStatus,
} from "../models/mesh.js";

/**
 * Per-call request signature accepted by every resource. Mirrors the
 * shared shape in `SeedClient.request` — kept narrow here so mesh stays
 * decoupled from `client.ts` imports.
 */
type RequestFn = <T>(
  method: string,
  path: string,
  opts?: CallOptions & {
    idempotent?: boolean;
  },
) => Promise<T>;

export interface MeshResource {
  /**
   * `GET /api/v1/network/mesh/status` — overlay health snapshot.
   *
   * Returns `{ ap_active, auto_mesh, connected_to_seed, device_id,
   * has_mesh_password, peer_count, peers }`. A single-seed install
   * reports `peer_count: 0, peers: []`.
   */
  status(opts?: CallOptions): Promise<MeshStatus>;

  /**
   * `GET /api/v1/peers` — list of known mesh peers as exposed by
   * the seed's discovery layer (`_cognitum._tcp.local` mDNS +
   * gossip). Returns `{ count, discovery_active, peers }`.
   */
  peers(opts?: CallOptions): Promise<MeshPeers>;

  /**
   * `GET /api/v1/swarm/status` — swarm coordination state (epoch,
   * peer count, local vector count snapshot).
   */
  swarmStatus(opts?: CallOptions): Promise<SwarmStatus>;

  /**
   * `GET /api/v1/cluster/health` — cluster-level aggregate.
   *
   * Returns `{ auto_sync_interval_secs, cluster_enabled,
   * discovery_active, last_sync_attempt, peer_count, peers }`.
   */
  clusterHealth(opts?: CallOptions): Promise<ClusterHealth>;
}

/** Build the mesh observability resource bound to `request`. */
export function makeMeshResource(request: RequestFn): MeshResource {
  return {
    status: (opts) =>
      request<MeshStatus>("GET", "/api/v1/network/mesh/status", {
        idempotent: true,
        ...(opts ?? {}),
      }),

    peers: (opts) =>
      request<MeshPeers>("GET", "/api/v1/peers", {
        idempotent: true,
        ...(opts ?? {}),
      }),

    swarmStatus: (opts) =>
      request<SwarmStatus>("GET", "/api/v1/swarm/status", {
        idempotent: true,
        ...(opts ?? {}),
      }),

    clusterHealth: (opts) =>
      request<ClusterHealth>("GET", "/api/v1/cluster/health", {
        idempotent: true,
        ...(opts ?? {}),
      }),
  };
}

File: sdks/rust/src/seed/resources/mesh.rs
//! Mesh observability resource — `/api/v1/{network/mesh/status,peers,
//! swarm/status,cluster/health}` (Phase 2 — ADR-0016a §D8).
//!
//! All four endpoints are allowlisted reads on firmware v0.20.0. Shapes
//! live in [`crate::seed::models::mesh`]; every struct carries a
//! `#[serde(flatten)] extras: Extras` so additional fields from newer
//! firmwares deserialize without breaking the SDK.

use crate::error::Error;
use crate::seed::client::SeedClient;
use crate::seed::config::CallOptions;
use crate::seed::models::{ClusterHealth, MeshPeers, MeshStatus, SwarmStatus};

/// Mesh observability endpoints.
pub struct MeshResource<'c> {
    pub(crate) client: &'c SeedClient,
}

impl<'c> MeshResource<'c> {
    /// `GET /api/v1/network/mesh/status` — mesh membership + AP snapshot.
    pub async fn status(&self) -> Result<MeshStatus, Error> {
        self.client.request_get("/network/mesh/status").await
    }

    /// [`Self::status`] with per-call [`CallOptions`] overrides.
    pub async fn status_with(&self, opts: CallOptions) -> Result<MeshStatus, Error> {
        self.client
            .request_get_opts("/network/mesh/status", &opts)
            .await
    }

    /// `GET /api/v1/peers` — discovery peer table.
    pub async fn peers(&self) -> Result<MeshPeers, Error> {
        self.client.request_get("/peers").await
    }

    /// [`Self::peers`] with per-call [`CallOptions`] overrides.
    pub async fn peers_with(&self, opts: CallOptions) -> Result<MeshPeers, Error> {
        self.client.request_get_opts("/peers", &opts).await
    }

    /// `GET /api/v1/swarm/status` — swarm-mode observability.
    pub async fn swarm_status(&self) -> Result<SwarmStatus, Error> {
        self.client.request_get("/swarm/status").await
    }

    /// [`Self::swarm_status`] with per-call [`CallOptions`] overrides.
    pub async fn swarm_status_with(&self, opts: CallOptions) -> Result<SwarmStatus, Error> {
        self.client.request_get_opts("/swarm/status", &opts).await
    }

    /// `GET /api/v1/cluster/health` — cluster / auto-sync status.
    pub async fn cluster_health(&self) -> Result<ClusterHealth, Error> {
        self.client.request_get("/cluster/health").await
    }

    /// [`Self::cluster_health`] with per-call [`CallOptions`] overrides.
    pub async fn cluster_health_with(&self, opts: CallOptions) -> Result<ClusterHealth, Error> {
        self.client.request_get_opts("/cluster/health", &opts).await
    }
}
